### PR TITLE
Scream yaml support

### PR DESCRIPTION
**Pull request by bartgol**
This PR adds yaml-cpp as a project-wide submodule (i.e., in externals, alongside kokkos and scorpio), and the utilities needed to parse a yaml file and fill a in-house `ParameterList` structure with the same data.

This will allow us to write a unit test once, and run it multiple times (and/or debug it) simply changing an input file.

### CODE REVIEW

**Comment by bartgol** [comment]
I'm confused by the failure on blake: I manually ran test-all-scream on blake, and everything passed. I haven't checked bowman. I am currently trying to figure out why yaml-cpp doesn't compile on white. If yaml-cpp ends up being intrinsically unusable on gpu platforms, I will try libyaml, a C library, slimmer, but harder to use (some work that yaml-cpp does for you, you have to manually do it with libyaml).

**Comment by ambrad** [comment]
Retest. Master should be clean now, which in turn should make your branch clean since AT will merge origin/master into your branch before testing.

**Comment by bartgol** [comment]
Thanks Andrew. I am still working on the gpu build, which is cause (per our dicussion offline) by yaml-cpp always adding the pedantic flag, which with nvcc causes issues.

I have opened an issue on their repo, offering to issue a PR if they agree that this can be a problem. However, I am going to fork yaml-cpp into e3sm, cause I'm not sure they will respond very fast. If/when they do, and if/when the pedantic flag gets removed, I will switch back the module to the upstream/master.

**Comment by bartgol** [comment]
Uhm, it seems that the new master did not solve the issue for the AT on blake/bowman. I'm not sure what's happening, I'm manually running test-all to then step into the test and debug a bit.

**Comment by bartgol** [comment]
Ok, I found out that running 
```./scream/components/scream/scripts/gather-all-data './scripts/test-all-scream $compiler -p -m $machine' -m blake
```
(which is what the jenkins job runs) yields a failure, like we see in the AT reports. However, running
```
cd scream/components/scream
./scripts/test-all-scream -m blake -i -p mpicxx
```
(with the same modules and env) works!

Notice that the first command does not perform the merge or origin/master (since I'm not adding the -i flag). On jenkins, the -i flag should be added automatically, because test-all detects we're in a jenkins job, so it does it. And in fact, in the jenkins jobs logs, I can see the line 
```
RUN: git merge origin/master
```
@jgfouca is it possible that, after the baseline generation, when test-all checks out the original branch, it doesn't get the merge commit in it? Maybe the lack of human interaction with the shel aborts the merge commit?

File: .gitmodules [hunk 1]
```diff
@@ -1,29 +1,32 @@
 [submodule "components/clm/src/external_models/fates"]
-	path = components/clm/src/external_models/fates
```
Comment (ambrad) [inline .gitmodules]: These whitespace changes will cause conflicts if the upstream changes parts of this file. IMO, it's best just to add your submodule and leave everything else as is.

Comment (bartgol) [reply .gitmodules]: Ah, it must have been my vim, changing tabs into spaces. But this does not explain why test-all would pass when I manually run it.

Comment (ambrad) [reply .gitmodules]: No, I mean the true upstream, E3SM. When we do periodic upstream/E3SM merges, it's good not to have to resolve conflicts in infrastructure files due to whitespace changes.

**Comment by bartgol** [comment]
So, I added a printout of git log right after switching back to the pr branch, and this is the output:
```
23:24:06   Switched back to pr/279 (2456cb1e704ba39463f043001a49ec4a252d31e0)
23:24:06   git log -1: commit f8df9783be82608bf96ad29e8da1c58eb5ec177d
23:24:06 Author: Luca Bertagna <lbertag@sandia.gov>
23:24:06 Date:   Sun Mar 8 23:02:48 2020 -0600
23:24:06 
23:24:06     SCREAM: verbosity mod to test-all-scream
23:24:06 Run dashboard with model Experimental
```
However, the last commit _should_ be the merge commit. I am suspecting something is off with the switch back/forth with master/branch, and with the merging of master. Notice that the sha in the first line does not match the one of the last commit though, which is another odd thing.

For clarity, these are the lines in test_all_scream.py that print the stuff above:

```
        if need_checkout:
            run_cmd_no_fail("git checkout {}".format(git_head_ref))
            print("  Switched back to {} ({})".format(git_head_ref, git_head_commit))
            log = run_cmd_no_fail("git log -1")
            print("  git log -1: {}".format(log))
```
So there are two odd things:

1. the sha of the branch printed on `switching back...` should match the sha of the commit printed by `git log -1`.
2. the last commit should be the merge of origin/master.

**Comment by bartgol** [comment]
Ah, I think I know the problem. The branch is `pr/279`, and _that's_ what git uses to switch back to the branch. However, that branch has _not_ been updated by the merge of master. So even if the sha looks ok, git switches back to the `pr/279` ref, rather than to the commit. This is a bug. Opening an issue.

File: .gitmodules (hunk 1 above)
Comment (bartgol) [reply .gitmodules]: Ah, I see. I will push a fix for this.

File: components/scream/cmake/YamlCpp.cmake [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                           submodule directory is only included once (with add_subdirectory).")
+
+GET_PROPERTY(IS_YAMLCPP_ALREADY_BUILT GLOBAL PROPERTY E3SM_YAMLCPP_ALREADY_BUILT SET)
+
+IF (NOT IS_YAMLCPP_ALREADY_BUILT)
+  SET (YAMLCPP_SRC_DIR ${SCREAM_SOURCE_DIR}/../../externals/yaml-cpp)
+  SET (YAMLCPP_BIN_DIR ${CMAKE_BINARY_DIR}/externals/yaml-cpp)
+
+  SET (BUILD_TESTING FALSE)
+  ADD_SUBDIRECTORY (${YAMLCPP_SRC_DIR} ${YAMLCPP_BIN_DIR})
```
Comment (jgfouca) [inline components/scream/cmake/YamlCpp.cmake]: This file is mixing case styles.

File: components/scream/src/physics/shoc/temp/atmosphere_macrophysics.cpp [hunk 3]
```diff
@@ -1,117 +0,0 @@
-#include "../../share/scream_assert.hpp"
-#include "physics/shoc/scream_shoc_interface.hpp"
-#include "physics/shoc/atmosphere_macrophysics.hpp"
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/temp/atmosphere_macrophysics.cpp]: Why were these files deleted?

File: components/scream/cmake/YamlCpp.cmake (hunk 2 above)
Comment (bartgol) [reply components/scream/cmake/YamlCpp.cmake]: Do you have any preference on which one to keep? I like upper case for cmake, but we used lower case in other places...

File: components/scream/src/physics/shoc/temp/atmosphere_macrophysics.cpp (hunk 3 above)
Comment (bartgol) [reply components/scream/src/physics/shoc/temp/atmosphere_macrophysics.cpp]: They were not supposed to be there to start with. I think Aaron told me he copied them in a temp dir while it modified the ones in physics/shoc, then forgot to delete the temp dir. Peter was going to delete that folder, then saw I would do it in a scream-ad pr, so he didn't. We can delete it in another pr, but I figured we may do it here, and save ourselves another round of AT/review.

To be clear, those files are not even built.

File: components/scream/cmake/YamlCpp.cmake (hunk 2 above)
Comment (jgfouca) [reply components/scream/cmake/YamlCpp.cmake]: all-caps looks awful to me for some reason. I prefer cmake-command to be lowercase, var names upper.

File: components/scream/src/physics/shoc/temp/atmosphere_macrophysics.cpp (hunk 3 above)
Comment (jgfouca) [reply components/scream/src/physics/shoc/temp/atmosphere_macrophysics.cpp]: OK, no problem
